design-document.md

```diff
@@ -14,4 +24,210 @@
+1. The joint randomness r is an element of K^u(n) for some function u.
+1. Each verification share {vf:i} is an element of K^v(n) for some function v.
+
+Before the protocol begins, the aggregators agree on joint randomness r and
+designate one of the aggregators as the leader. The protocol proceeds as
+follows:
+
+1. The user runs pf := Prove(x). It splits x and pf into {x} and {pf}
+   respectively and sends ({x:i}, {pf:i}) to aggregator i.
+1. Each aggregator i runs {vf:i} := Query({x:i}, {pf:i}, r) ands sends {vf:i} to
+   the leader.
+1. The leader runs b := Decide({vf}, r) and sends b to each of the aggregators.
+
+If b=1, then each aggregator i adds its input share {x:i} into its share of the
+aggregate. Once all of the inputs have been validated and aggregated, the
+aggregators send their aggregate shares to the leader, who adds them together to
+obtain the final result.
+
+[[TODO: Sketch out the b=1 path.]]
+
+**Proof generation and verification.**
+[[TODO: Describe how to construct proof systems for languages recognized by
+validity circuits with G-gates, a la [BCC+19, Theorem 4.3].]]
+
+**Security parameters.**
+[[TODO: Define completeness, soundness, and honest-verifier zero-knowledge for
+fully linear PCPs and state bounds for [BBC+19, Theorem 4.3]. This bound will
+guide the selection of the field best suited for the data type and
+application.]]
+
+**Consensus protocol.**
+[[TODO: Describe how the aggregators pick the leader and the joint randomness.]]
+
+**Key distribution.**
+[[TODO: Decide how users obtain aggregators' public keys.]]
+
+### Changes to the input-validation protocol
+
+**Coordinating state.**
+The state of the input-validation protocol is maintained by the leader; except
+for aggregation of the input shares, the other aggregators are completely
+stateless. This is accomplished by making the following changes to the core
+protocol.
+
+1. The user sends all of its shares to the leader. To maintain privacy, the user
+   encrypts each (input, proof) share under the public key of the share's
+   recipient.
+1. The leader forwards each encrypted share to its intended recipient. Each
+   aggregator decrypts its input and proof share, computes its verification
+   share, and sends its verification share to the aggregator as usual.
+1. If b=1 in the last step, then the leader also sends along the encrypted input
+   share to each aggregator so that they can decrypt and aggregate the share
+   without needing to cache the input share from the previous step.
+
+**Minimizing bandwidth overhead.**
+In most linear secret sharing schemes, the length of each share is equal to the
+length of the input. Therefore, the communication overhead for the user is
+O(s\*(n+p(n))). This can be reduced to O(s+n+p(n)) with the following standard
+trick.
+
+Let x be an element of K^n for some n. Suppose we split in into {x} by choosing
+{x:1}, ..., {x:s-1} at random and letting {x:s} = in - ({x:1} + ... + {x:s-1}).
```

Comment:
I'm a bit confused by `in` vs. `x` in this paragraph. Is `in` meant to be the input before being split into the vector `{x}`? If so, should line 116 read "Let in be an element of..."?

Reply:
Oops, typo: s/in/x/.